0.md

```diff
@@ -0,0 +1,322 @@
+text.
+A `Module` instance represents the lifecycle of a EcmaScript module and allows
+virtual import behavior.
+Multiple `Module` instances can share a common `ModuleSource`.
+
+## Interfaces
+
+### ModuleSource
+
+```ts
+interface ModuleSource {
+  constructor(source: string);
+}
+```
+
+Semantics: A `ModuleSource` instance gives the holder no powers.
+It represents a compiled EcmaScript module and does not capture any information
+beyond what can be inferred from a module's source text.
+Import specifiers in a module's source text cannot be interpreted without
+further information.
+
+Note 1: `ModuleSource` does for modules what `eval` already does for scripts.
+We expect content security policies to treat module sources similarly.
+A `ModuleSource` instance constructed from text would not have an associated
+origin.
+A `ModuleSource` instance can be constructed from vetted text and host-defined
+import hooks may reveal module sources that were vetted behind the scenes.
+
+Note 2: Multiple `Module` instances can be constructed from a single `ModuleSource`,
+producing one exports namespaces for each imported `Module` instance.
+
+Note 3: The internal record of a `ModuleSource` instance is immutable and
+serializable.
+This data can be shared without cost between realms of an agent or even agents
+of an agent cluster.
+
+### Module instances
+
+```ts
+type ImportSpecifier = string;
+
+type ImportHook = (specifier: ImportSpecifier, importMeta: object) =>
+  Promise<Module>;
+
+interface Module {
+  constructor(
+    source: ModuleSource,
+    options: {
+      importHook?: ImportHook,
+      importMeta?: object,
+    },
+  );
+
+  readonly source: ModuleSource,
+}
+```
+
+Semantics: A `Module` has a 1-1-1-1 relationship with a ***Module Environment
+Record***, a ***Module Record*** and a ***module namespace exotic object***.
+
+The module has a lifecycle and fresh instances have not been linked,
+initialized, or executed.
+
+Invoking dynamic import on a `Module` instances advances it and its transitive
+dependencies to their end state.
+Consistent with dynamic import for a stringly-named module,
+dynamic import on a `Module` instance produces a promise for the corresponding
+***Module Namespace Object***
+
+Dynamic import induces calls to `importHook` for each unsatisfied dependency of
+each module instance in separate events, before any dependency advances to the
+link phase of its lifecycle.
+
+Dynamic import within the evaluation of a `Module` also invokes the
+`importHook`.
+
+`Module` instances memoize the result of their `importHook` keyed on the given
+Import Specifier.
+
+`Module` constructors like `Function` constructors are bound to a realm
+and evaluate modules in their particular realm.
+
+## Examples
+
+### Import Kicker
+
+Any dynamic import function is suitable for initializing, linking, and
+evaluating a module instance and all of its transitive dependencies.
```

Comment:
Given that the immediate `importHook` returns a (promise for a) module instance, and that module instance presumably has its own importHook, how does the transitive use of the original importHook arise *at this level*? I suspect it does not and this "transitive" explanation should be omitted *at this level*.

Reply:
Import hooks do produce shallow `Module` instances and import hooks may return module instances in any state of their lifecycle.

This section describes dynamic import, which guarantees that the given module instance is advanced to its terminal state. I’m not sure whether it’s possible to advance a module to its terminal state without advancing its transitive dependencies to their terminal state too.

Reply:
> how does the transitive use of the original importHook arise at this level?

I believe (and I implemented it as) it does not "transitive". When resolving modules, it will call the parent module's importHook, and every resolved module `T` will have its own importHook to resolve `T`'s dependencies, so the importHook only affects 1 level.

Reply:
Yes, each `Module` instance uses its own `importHook`. Dynamic import asynchronously advances the target `Module` instance to its terminal state. That implies advancing its transitive dependencies, either all to their terminal success state, or any one of them to its terminal failure state.